### PR TITLE
fix(windows): avoid cached hotkey state

### LINKED ISSUE #4884: bug(windows): modifier keys sometimes get stuck
**Describe the bug**

https://community.software.sil.org/t/problems-with-keyman-windows-popping-up/4668

In some situations Keyman thinks that modifier keys are down for hotkeys when in fact they are not. This modifier state does not impact other input, just hotkeys within Keyman such as open osk.

Workaround: press and release each modifier to reset the hotkey modifier state.

---

**Keyman for Windows/macOS/Linux:**
 - OS: Windows 10 
 - Keyman version: 14.0.271 
 - Target Application:

### PR DESCRIPTION
**Pull request by mcdurdin**
Fixes #4884.

This adds a new flag `Flag_UseCachedHotkeyModifierState` (`engine.compatibility.old_cached_hotkey_modifier_state` in Settings app), to allow you to turn this setting off again.

### CODE REVIEW

File: windows/src/engine/keyman32/k32_lowlevelkeyboardhook.cpp [hunk 1]
```diff
@@ -145,7 +155,18 @@ LRESULT _kmnLowLevelKeyboardProc(
   SendDebugMessageFormat(0, sdmAIDefault, 0, "kmnLowLevelKeyboardProc: wparam: %x  lparam: %x [vk:%s scan:%x flags:%x extra:%x]", wParam, lParam, Debug_VirtualKey((WORD) hs->vkCode), hs->scanCode, hs->flags, hs->dwExtraInfo);   // I4674
 
   DWORD Flag = 0;
-  if (UseRegisterHotkey()) {
+  if (!UseCachedHotkeyModifierState()) {
+    // #5190: Don't cache modifier state because sometimes we won't receive
+    // modifier change events (e.g. on lock screen)
+    FHotkeyShiftState = 0;
+    if (GetKeyState(VK_LCONTROL) < 0) FHotkeyShiftState |= HK_CTRL;
```
Comment (darcywong00) [inline windows/src/engine/keyman32/k32_lowlevelkeyboardhook.cpp:162]: ll. 184-185 says 
> The right modifiers are not used for hotkeys...

Does that mean VK_RCONTROL, VK_RMENU, and VK_RSHIFT can be ignored?

File: windows/src/global/delphi/general/Keyman.System.Settings.pas [hunk 2]
```diff
@@ -396,6 +397,18 @@   TKeymanSettings = class(TObjectList<TKeymanSetting>)
...
     ),
 
+    // keyman32:k32_lowlevelkeyboardhook
+    (
+      ID: 'engine.compatibility.old_cached_hotkey_modifier_state';
+      Name: SRegValue_Flag_UseCachedHotkeyModifierState;
+      RootKey: HKCU;
+      Key: SRegKey_KeymanEngineDebug_CU;
+      Description: 'Set to 1 to use a cached modifer state when checking hotkeys; if '+
+                   'set to0 then engine.compatibility.old_hotkey_registration may not '+
```
Comment (darcywong00) [inline windows/src/global/delphi/general/Keyman.System.Settings.pas:407]:
```suggestion
                   'set to 0 then engine.compatibility.old_hotkey_registration may not '+
```

File: windows/src/engine/keyman32/k32_lowlevelkeyboardhook.cpp (hunk 1 above)
Comment (mcdurdin) [reply windows/src/engine/keyman32/k32_lowlevelkeyboardhook.cpp:162]: No, because if the user is pressing them then we don't want to activate the hotkey.

**Review by darcywong00 — APPROVED** [review]
lgtm

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 15.0.59-alpha](https://keyman.com/downloads/releases/alpha/15.0.59)

**Pull request merged**